### PR TITLE
Enable `Add-Content` to share read access to other tools while writing content

### LINKED ISSUE #5924: Why does Add-Content require a read lock when appending to files?
I (and various customers of my company) notice that `Add-Content` often fails to append to files with an error like:

```
Add-Content : The process cannot access the file
'C:\Users\Administrator\logs\access_log.log' because it is being used by another process.
```

This happens if the file in question is being read ("tailed") by some other process at the same time, e.g. a log collection framework.

I confirmed via ProcMon that this is because `Add-Content` takes a read lock on its target file. So if the file is already open for read by another process, the cmdlet will fail.

This is unexpected. Why does a **writing** cmdlet need a **read** lock? `Out-File -Append` does essentially the same thing and takes no such lock. Others are confused as well: https://stackoverflow.com/questions/30733610/get-content-wait-is-locking-the-file-while-it-reads-it-preventing-add-content, https://stackoverflow.com/questions/10655788/powershell-set-content-and-out-file-what-is-the-difference

As far as I can surmise the relevant bit of code is here (`FileSystemProvider.GetContentWriter`):

https://github.com/PowerShell/PowerShell/blob/4bc52d2358222084738157a08907fac32d13bd3a/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L6808

Could `FileShare.Write` be changed to `FileShare.ReadWrite`?

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
## PR Summary

`Add-Content` uses a writer that currently sets share access to write.  So if using a tool like `tail` to read a log file while trying to write to it will fail.  Fix is to change share access to readwrite which will allow other processes to continue to read from it while the writer writes new content.

Fix https://github.com/PowerShell/PowerShell/issues/5924

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1 [hunk 1]
```diff
@@ -59,3 +59,37 @@ Describe "Add-Content cmdlet tests" -Tags "CI" {
     }
   }
 }
+
+Describe "Add-Content feature tests" -Tag Feature {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1]: We always have problems with async tests. Do we really need the test? I'd only add an comment in C# code.

Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1]: I kind of agree on async tests being inherently fragile.  Perhaps for this one a comment in the code is sufficient.

Comment (anmenaga) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1]: Make it sync:
1) open a file with `FileStream(FileAccess.Read, FileShare.ReadWrite)` and keep it open.
2) run new `Add-Content` (with `FileAccess.Write, FileShare.ReadWrite`)
3) close `FileStream` in 1). The test should not throw.

The old `Add-Content` (with `FileShare.Write`) in 2) would throw.

**Review by anmenaga — APPROVED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1 (hunk 1 above)
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1]: Good suggestion, let me see if that repros the original issue.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1 [hunk 2]
```diff
@@ -59,3 +59,37 @@ Describe "Add-Content cmdlet tests" -Tags "CI" {
...
+  It "Should not block reads while writing" {
+    $logpath = Join-Path $testdrive "test.log"
+
+    $writer = Start-Job -ScriptBlock {
+      param($logpath)
+
+      1..10 | % {
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1]: Use `ForEach-Object` instead of `%`

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1 (hunk 1 above)
Comment (adityapatwardhan) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1]: Please see if these changes are possible. It would harden the test.

Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1]: @anmenaga that worked perfectly and so simple :)

**Review by adityapatwardhan — APPROVED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1 (hunk 2 above)
Comment (voytas75) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Add-Content.Tests.ps1]: is `(1..10).foreach{...}` faster then `ForEach-Object`?

**Pull request merged**